R/buttons.R

```diff
@@ -0,0 +1,292 @@
+      if (!is_task_button_manual_reset(shinysession, name)) {
+        update_task_button(name, state = "ready", session = shinysession)
+      }
+    })
+  }
+
+  # mark up the action button value with a special class so we can recognize it later
+  class(value) <- c("shinyActionButtonValue", class(value))
+  value
+}
+
+#' Bind `input_task_button` to `ExtendedTask`
+#'
+#' @description
+#' Sets up a [shiny::ExtendedTask] to relay its state to an existing
+#' [input_task_button()], so the task button stays in its "busy" state for as
+#' long as the extended task is running.
+#'
+#' Note that `bind_task_button` does _not_ automatically cause button presses to
+#' invoke the extended task; you still need to use [shiny::bindEvent()] (or
+#' [shiny::observeEvent()]) to cause the button press to trigger an invocation,
+#' as in the example below.
+#'
+#' @param target The target object (i.e. `ExtendedTask`).
+#' @param task_button_id A string matching the `id` argument passed to the
+#'   corresponding [input_task_button()] call.
+#' @param session A Shiny session object (the default should almost always be
+#'   used).
+#' @param ... Further arguments passed to other methods.
+#'
+#' @returns The `target` object that was passed in.
+#'
+#' @examplesIf rlang::is_interactive()
+#'
+#' library(shiny)
+#' library(bslib)
+#' library(future)
+#' plan(multisession)
+#'
+#' ui <- page_sidebar(
+#'   sidebar = sidebar(
+#'     input_task_button("recalc", "Recalculate")
+#'   ),
+#'   textOutput("outval")
+#' )
+#'
+#' server <- function(input, output) {
+#'   rand_task <- ExtendedTask$new(function() {
+#'     future({
+#'       # Slow operation goes here
+#'       Sys.sleep(2)
+#'       runif(1)
+#'     }, seed = TRUE)
+#'   }) |> bind_task_button("recalc") # Make button state reflect task
+#'
+#'   observeEvent(input$recalc, {
+#'     rand_task$invoke()
+#'   })
+#'
+#'   output$outval <- renderText({
+#'     rand_task$result()
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+#'
+#' @export
+bind_task_button <- function(target, task_button_id, ...) {
+  UseMethod("bind_task_button")
+}
+
+#' @rdname bind_task_button
+#' @export
+bind_task_button.default <- function(target, task_button_id, ...) {
+  abort(
+    "Don't know how to bind a task button to an object of class '",
+    class(target)[[0]],
+    "'"
+  )
+}
+
+#' @rdname bind_task_button
+#' @export
+bind_task_button.ExtendedTask <- function(target, task_button_id,
+  ..., session = get_current_session()) {
+
+  force(target)
+  force(task_button_id)
+  force(session)
+
+  was_running <- FALSE
+  shiny::observe({
+    running <- target$status() == "running"
+    if (running != was_running) {
+      was_running <<- running
+      if (running) {
+        update_task_button(task_button_id, state = "busy", session = session)
+      } else {
+        update_task_button(task_button_id, state = "ready", session = session)
+      }
+    }
+  }, priority = 1000)
+  return(target)
+}
```

Comment:
Would it be reasonable to expect that, if a button is bound to multiple tasks, that the button should stay in progress until _all_ tasks are complete? Or should we say that you need `auto_reset = FALSE` for that case?

```r
library(shiny)
library(bslib)
library(future)
library(promises)

future::plan(future::multisession)

ui <- page_sidebar(
  sidebar = sidebar(
    numericInput("x", "x", 5),
    numericInput("y", "y", 6),
    input_task_button("calc", "Calculate")
  ),
  verbatimTextOutput("status"),
  verbatimTextOutput("results", placeholder = TRUE)
)

server <- function(input, output, session) {
  t1 <- ExtendedTask$new(function(x, y) {
    future_promise({
      Sys.sleep(1)
      x + y
    })
  }) |> bind_task_button("calc")
  
  t2 <- ExtendedTask$new(function(x, y) {
    future_promise({
      Sys.sleep(5)
      x + y
    })
  }) |> bind_task_button("calc")
  
  observeEvent(input$calc, {
    t1$invoke(input$x, input$y)
    t2$invoke(input$x + 1, input$y + 1)
  })
  
  output$status <- renderText({
    paste(t1$status(), "-", t2$status())
  })
  
  output$results <- renderText({
    paste(t1$result(), "-", t2$result())
  })

}

shinyApp(ui, server)
```

Reply:
Hmmm... good question. I didn't think about this at all. I think for now we'd need to say "don't do that"?

Workaround is, what? `auto_reset=FALSE`, and then:
```
observe({
  if (t1$status() != "running" && t2$status() != "running") {
    update_task_button("calc", state = "ready")
  } else {
    update_task_button("calc", state = "busy")
  }
})
```

Reply:
I added a note to the docs that this isn't supported